### ISSUE TITLE
Rule Request: Prefer specialized collection types over Array literals

### ISSUE THREAD

**Issue by lukaskollmer** [realm/SwiftLint#6152]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New Rule Request

Please describe the rule idea, format this issue's title as `Rule Request: [Rule Name]` and describe:

The idea here is for a rule, called "Prefer specialized collection types over Array literals", which would trigger in the following two cases:
1. when calling a function that takes a generic `Collection` parameter, but passing in an `Array` literal containing only a single element.
2. when calling a function that takes a generic `Sequence` or `Collection` parameter, but passing in an empty `Array` literal.

Why should this rule be added?
- it's primarily a performance rule, since using these specialized collection types allows avoiding having to allocate and initialize an Array instead.

Examples that would / would not trigger the rule:
```swift
func f1(_ elements: some Collection<String>) { ... }
func f2<S: Sequence>(_ elements: S) { ... }
func f3(_ elements: [String]) { ... }

f1([""]) // would trigger, suggesting that `CollectionOfOne("")` be used instead
f2([""]) // would trigger, suggesting that `CollectionOfOne("")` be used instead
f3([""]) // would not trigger

f1(["", ""]) // would not trigger
f2(["", ""]) // would not trigger
f3(["", ""]) // would not trigger

f1([]) // would trigger, suggesting that `EmptyCollection()` be used instead
f2([]) // would trigger, suggesting that `EmptyCollection()` be used instead
f3([]) // would not trigger
```

Should the rule be configurable?
- i guess it'd make sense to allow selecting that you want only one of the `EmptyCollection` or `CollectionOfOne` suggestions to happen.

Should the rule be opt-in or out-out?
- it should probably be opt-in, i.e. disabled by default.
